tsdb/querier.go

```diff
@@ -239,13 +239,17 @@ func PostingsForMatchers(ctx context.Context, ix IndexReader, ms ...*labels.Matc
 			return nil, ctx.Err()
 		}
 		switch {
-		case m.Name == "" && m.Value == "": // Special-case for AllPostings, used in tests at least.
+		case m.Name == "" && m.Value == "" || m.Type == labels.MatchRegexp && (m.Value == ".*" || m.Value == "^.*$"):
+			// Special-case for AllPostings
```

Comment:
Do we really want to go this way - adding every series in the index and then filtering it down?

That seems a lot of work, when we could mostly just ignore `=~".*"`.

Reply:
You are right, and numbers are better:
```
BenchmarkQuerier/Head/PostingsForMatchers/n="1",_j="foo"-8                  1511605               746.8 ns/op
BenchmarkQuerier/Head/PostingsForMatchers/n="1",_i=~".*",_j="foo"-8                 1561862               827.1 ns/op

```

Updated in last commit